lib/errors/error_handling.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require_dependency "#{Rails.root}/lib/errors/course_update_error_handling"
+
+module ErrorHandling
+  include CourseUpdateErrorHandling
+
+  def perform_error_handling(error, typical_errors, extra, course, optional_params)
+    Rails.logger.info "Caught #{error}"
+    level = error_level(error, typical_errors)
+    if course.present?
+      perform_course_error_handling(error, level, extra, course, optional_params)
+    else
+      report_exception_sentry(error, level, extra)
+    end
+    return nil
+  end
+
+  def error_level(error, typical_errors)
+    typical_errors.include?(error.class) ? 'warning' : 'error'
+  end
+
+  def raise_unexpected_error(error, typical_errors)
+    raise error if typical_errors.exclude?(error.class)
+  end
+
+  def report_exception_sentry(error, level, extra)
+    Raven.capture_exception error, level: level, extra: extra
+  end
+
+  # Building optional params for Replica and OresApi which require
+  # response body for Oj::ParseError to be saved in CourseErrorRecord
+  # WikiApi does not have Oj::ParseError as a typical error, which is
+  # why it does not use this method
+  def build_optional_params(course, error, url, response_body)
```

Comment:
I don't like coupling this to the two classes that would use "optional_params"; it means that ErrorHandling and Replica each have to "know something" about the internals of the other.

Instead, I recommend making all the data needed for this (url, response body) into optional named arguments of `perform_error_handling`, so that the only thing other classes need to "know" is the method signature of `perform_error_handling`.

Reply:
In `perform_error_handling`, the number of arguments were getting out of hand(around 7), so what I did was added an attribute `optional_params[:build]`, passed on `optional_params` to `CourseUpdateErrorHandling` module and then executed `build_optional_params` if `optional_params[:build]` was `true`. Also, I moved the method `build_optional_params` to the relevant module `CourseUpdateErrorHandling` and called that method in `save_course_course_record` where it was ultimately required.